### PR TITLE
Updates and fixes for Akka.IO UDP support

### PR DESCRIPTION
**Pull request by Horusiath**
This should fix the issue reported in #2995 (PR contains tests which reproduced the issue). It also upgrades the codebase to C# 7.

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
@Horusiath some UDP specs on Mono failing following this change

**Comment by Danthar** [comment]
Because you also refactored to use new patterns and stuff. Its hard to review. Could you point out what you changed to fix the problem ?

**Comment by Horusiath** [comment]
Essentially the problem was that I didn't updated the UDP handler code in line to the changes I was making in `TcpConnection` during Akka.IO rework. In previous versions we were sending `ByteString` via socket chunk by chunk (waiting for the previous chunk to be sent before sending next one). New implementation (that already is present in `TcpConnection`) makes smart distinction between compact and multi-chunk `ByteString`s:

- Compact byte string consist only of single chunk, so it's simply set as `SocketAsyncEventArgs.SetBuffer(buffer.Array, buffer.Offset, buffer.Count)` before sending.
- However multi-chunk byte string can also be send in a single call, thanks to `SocketAsyncEventArgs.BufferList` property.

I've extracted logic responsible for recognizing which method to use (`SetBuffer`/`BufferList`) from `TcpConnection` into single extension method [SocketAsyncEventArgs.SetBuffer(ByteString)](https://github.com/akkadotnet/akka.net/pull/3005/files#diff-c1896e208456b262362b837e0f76cd05), which now allows us to handle socket send request via TCP/UDP in quite similar way.

**Comment by alexvaluyskiy** [comment]
@Horusiath you made a lot of API changes, it will be impossible to ship it in 1.3.x. Is it possible to bring back some constructors and properties and make them obsoleted?

**Comment by Horusiath** [comment]
@alexvaluyskiy 90% of API changes is about sealing message classes (which should be sealed). There's a one change in the `Send` constructor, but the problem is more that you shouldn't even build it via constructor - which should be hidden - but rather via `Send.Create(ByteString)`. I can make an obsoleted constructor, but `Send.Payload` should be `ByteString` anyway.

**Comment by Horusiath** [comment]
The reason of failure on Mono found: *Mono doesn't support using BufferList at this point.* This probably means, that the same problem concerns TCP, we just didn't have tests to detect that. 

As I definitely don't want to fallback the whole ability to send multiple buffers at once or build a separate (and very complex) buffer send awaiting logic just to cover Mono, my proposal is: If we detect, that we're on Mono, we use `ByteString.Compact()` method, to squash multi-buffer `ByteString` into single-buffer one. This is copying operation. it sucks for performance and memory usage, but it will cripple only some of the cases (multi-buffer `ByteString`s) and only on Mono.

Are we fine with that? /cc @akkadotnet/core

File: src/core/Akka/IO/SocketEventArgsPool.cs [hunk 1]
```diff
@@ -80,4 +81,38 @@ private SocketAsyncEventArgs CreateSocketAsyncEventArgs()
...
+                args.SetBuffer(buffer.Array, buffer.Offset, buffer.Count);
+            }
+            else
+            {
+                if (RuntimeDetector.IsMono)
+                {
+                    // Mono doesn't support BufferList - falback to compacting ByteString
+                    var compacted = data.Compact();
+                    var buffer = compacted.Buffers[0];
+                    args.SetBuffer(buffer.Array, buffer.Offset, buffer.Count);
```
Comment (Horusiath) [inline src/core/Akka/IO/SocketEventArgsPool.cs]: If we detect a Mono runtime, we'll try to fallback and compact multiple byte buffers into one (this operations copies buffers onto newly allocated array).

**Comment by Aaronontheweb** [comment]
> If we detect, that we're on Mono, we use ByteString.Compact() method, to squash multi-buffer ByteString into single-buffer one. 

Fine with me. Not our job to implement the missing parts of the Mono runtime.

File: src/core/Akka.Tests/IO/UdpIntegrationSpec.cs [hunk 2]
```diff
@@ -61,6 +62,20 @@ public void The_UDP_Fire_and_Forget_implementation_must_be_able_to_send_without_
         }
 
         [Fact]
+        public void The_UDP_Fire_and_Forget_implementation_must_be_able_to_send_multipart_ByteString_without_binding()
```
Comment (Aaronontheweb) [inline src/core/Akka.Tests/IO/UdpIntegrationSpec.cs]: Looks good to me

File: src/core/Akka/IO/SocketEventArgsPool.cs (hunk 1 above)
Comment (Aaronontheweb) [inline src/core/Akka/IO/SocketEventArgsPool.cs]: looks good to me

File: src/core/Akka/IO/UdpConnected.cs [hunk 3]
```diff
@@ -152,44 +152,48 @@ public NoAck(object token)
         /// </summary>
         public sealed class Send : Command
         {
+            [Obsolete("Akka.IO.Udp.Send public constructors are obsolete. Use `Send.Create` or `Send(ByteString, EndPoint, Event)` instead.")]
+            public Send(IEnumerator<ByteBuffer> payload, Event ack)
```
Comment (Aaronontheweb) [inline src/core/Akka/IO/UdpConnected.cs]: Looks good

File: src/core/Akka/Util/ByteString.cs [hunk 4]
```diff
@@ -35,6 +35,16 @@ public sealed class ByteString : IEquatable<ByteString>, IEnumerable<byte>
         #region creation methods
 
         /// <summary>
+        /// INTERNAL API: remove this method once <see cref="Udp.Send"/> public constructor will be removed.
+        /// </summary>
+        internal static ByteString FromBuffers(IEnumerator<ByteBuffer> buffers)
```
Comment (Aaronontheweb) [inline src/core/Akka/Util/ByteString.cs]: Looks good

**Pull request merged**